sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1695,16 +1695,176 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
+   * the Aggregate (second part of returned value)
+   */
+  private def splitSubquery(plan : LogicalPlan) : Tuple2[Seq[LogicalPlan], Aggregate] = {
+    var topPart = List[LogicalPlan]()
+    var bottomPart : LogicalPlan = plan
+    while (! bottomPart.isInstanceOf[Aggregate]) {
+      topPart = bottomPart :: topPart
+      bottomPart = bottomPart.children.head
+    }
+    (topPart, bottomPart.asInstanceOf[Aggregate])
+  }
+
+  /**
+   * Rewrite the nodes above the Aggregate in a subquery so that they generate an
+   * auxiliary column "isFiltered"
+   * @param subqueryPlan plan before rewrite
+   * @param filteredId expression ID for the "isFiltered" column
+   */
+  private def addIsFiltered(subqueryPlan : LogicalPlan, filteredId : ExprId) : LogicalPlan = {
+    val isFilteredRef = AttributeReference("isFiltered", BooleanType)(exprId = filteredId)
+    val (topPart, aggNode) = splitSubquery(subqueryPlan)
+    var rewrittenQuery: LogicalPlan = null
+    if (topPart.size > 0 && topPart.head.isInstanceOf[Filter]) {
+      // Correlated subquery has a HAVING clause
+      // Rewrite the Filter into a Project that returns the value of the filtering predicate
+      val origFilter = topPart.head.asInstanceOf[Filter]
+      var topRemainder = topPart.tail
+      val newProjectList =
+        origFilter.output :+ Alias(origFilter.condition, "isFiltered")(exprId = filteredId)
+      val filterAsProject = Project(newProjectList, origFilter.child)
+
+      rewrittenQuery = filterAsProject
+      while (topRemainder.size > 0) {
+        rewrittenQuery = topRemainder.head match {
+          case Project(origList, _) => Project(origList :+ isFilteredRef, rewrittenQuery)
+          case SubqueryAlias(alias, _) => SubqueryAlias(alias, rewrittenQuery)
+        }
+        topRemainder = topRemainder.tail
+      }
+    } else {
+      // Correlated subquery without HAVING clause
+      // Add an additional Project that adds a constant value for "isFiltered"
+      rewrittenQuery = Project(subqueryPlan.output :+ Alias(Literal(false), "isFiltered")
+      (exprId = filteredId), subqueryPlan)
+    }
+    return rewrittenQuery
+  }
+
+  /**
    * Construct a new child plan by left joining the given subqueries to a base plan.
    */
   private def constructLeftJoins(
       child: LogicalPlan,
       subqueries: ArrayBuffer[ScalarSubquery]): LogicalPlan = {
     subqueries.foldLeft(child) {
       case (currentChild, ScalarSubquery(query, conditions, _)) =>
-        Project(
-          currentChild.output :+ query.output.head,
-          Join(currentChild, query, LeftOuter, conditions.reduceOption(And)))
+        val origOutput = query.output.head
+
+        val resultWithZeroTups = evalSubqueryOnZeroTups(query)
+        if (resultWithZeroTups.isEmpty) {
+          Project(
+            currentChild.output :+ origOutput,
+            Join(currentChild, query, LeftOuter, conditions.reduceOption(And)))
+        } else {
```

Comment:
@hvanhovell the conditions for the COUNT bug in this rule are actually a bit less strict. If the correlated subquery returns an answer other than NULL when the correlation bindings do not join, then it may be possible to have the COUNT bug. Take this query for example:
```sql
select l.a from l where
        (select case when count(*) = 1 then null else count(*) end as cnt
        from r where l.a = r.c) = 0
```
This subquery returns NULL when exactly 1 tuple from `r` joins with a set of correlation bindings and a non-NULL value otherwise. Rewrite this query without compensating for the COUNT bug and you get:
```sql
select l.a from l left join 
  (select r.c, case when count(*) = 1 then null else count(*) end as cnt from r group by r.c) SQ
on l.a = SQ.c
where SQ.cnt = 0
```
If the value of `l.a` matches zero tuples in `r`, the outer join produces a tuple with `cnt` set to NULL and the rewritten query incorrectly returns an empty result.

Technically the full condition for the COUNT bug is that the subquery returns a non-NULL value when zero tuples join; *and* the outer query block will return a different answer if it receives that value instead of NULL. Checking the outer query block would be a somewhat-useful performance optimization, but it would require additional static evaluation code.